File: crawl/src/crwPredict.f90
SUBROUTINE crw_predict( tau2y, tau2x, Qmat, Tmat, x, y, loctype, &
                             stay, ay, ax, Py, Px, lonadj, N, lly, llx, &
                             My, uy, jky, Mx, ux, jkx, predy, predx, vary, varx)
 !! LOADED VALUES/ARRAYS !!
  INTEGER           N
  INTEGER           loctype(N), stay(N)
  DOUBLE PRECISION  lly, llx, tau2y(N), tau2x(N), Qmat(N,3), Tmat(N,2)
  DOUBLE PRECISION  x(N), y(N), lonadj(N)
  DOUBLE PRECISION  uy(N,1), ux(N,1), jky(N), jkx(N), My(N,1), Mx(N,1)
  DOUBLE PRECISION  ax(2,1), ay(2,1), Py(2,2), Px(2,2)
  DOUBLE PRECISION  predy(N,2), predx(N,2), vary(2,2,N), varx(2,2,N)

 !! DERIVED ARRAYS !!
  DOUBLE PRECISION vy(N), vx(N), Fy(N), Fx(N)
  DOUBLE PRECISION Z(1,2)
  DOUBLE PRECISION KyArr(2,1,N), KxArr(2,1,N)
  DOUBLE PRECISION Qy(2,2), Qx(2,2), T(2,2)
  DOUBLE PRECISION PyArr(2,2,N+1), PxArr(2,2,N+1), LyArr(2,2,N), LxArr(2,2,N)
  DOUBLE PRECISION ayArr(2,1,N+1), axArr(2,1,N+1)
  DOUBLE PRECISION ry(2,1), Ny(2,2), rx(2,1), Nx(2,2)

 !! INTIAL CONDITIONS !!
  ayArr(:,:,1) = ay
  axArr(:,:,1) = ax
  PyArr(:,:,1) = Py
  PxArr(:,:,1) = Px
  Z            = RESHAPE((/1.0, 0.0/),(/1, 2/))
  Qy           = 0.0
  Qx           = 0.0
  T            = 0.0
  T(1,1)       = 1.0
  vy           = 0.0
  vx           = 0.0
  Fy           = tau2y
  Fx           = tau2x
  ry           = 0.0
  rx           = 0.0
  Ny           = 0.0
  Nx           = 0.0

 !! BEGIN FILTER LOOP !!
  DO i=1,N

   !! GENERATE Q AND T MATRICES !!
    IF(stay(i)==1) THEN
      Qy = 0.0
      Qx = 0.0
      T(1,2) = 0.0
      T(2,2) = 0.0
    ELSE
      Qy(1,1) = Qmat(i,1)
      Qy(2,1) = Qmat(i,2)
      Qy(1,2) = Qy(2,1)
      Qy(2,2) = Qmat(i,3)
      Qx = Qy/(lonadj(i)*lonadj(i))
      T(1,2) = Tmat(i,1)
      T(2,2) = Tmat(i,2)
    END IF

   !! GENERAL KF FILTER !!
    Fy(i) = PyArr(1,1,i) + tau2y(i)
    Fx(i) = PxArr(1,1,i) + tau2x(i)/(lonadj(i)*lonadj(i))
    IF(loctype(i)==1 .OR. Fy(i)==0.0) THEN
      ayArr(:,:,i+1) = MATMUL(T,ayArr(:,:,i))
      PyArr(:,:,i+1) = MATMUL(MATMUL(T,PyArr(:,:,i)),TRANSPOSE(T)) + Qy
      KyArr(:,:,i) = 0.0
      LyArr(:,:,i) = T
    ELSE
      vy(i) = y(i)-ayArr(1,1,i)
      lly = lly - (log(Fy(i)) + vy(i)*vy(i)/Fy(i))/2
      KyArr(:,:,i) = MATMUL(MATMUL(T,PyArr(:,:,i)),TRANSPOSE(Z))/Fy(i)
      LyArr(:,:,i) = T - MATMUL(KyArr(:,:,i),Z)
      ayArr(:,:,i+1) = MATMUL(T,ayArr(:,:,i)) + KyArr(:,:,i)*vy(i)
      PyArr(:,:,i+1) = MATMUL(MATMUL(T,PyArr(:,:,i)),TRANSPOSE(LyArr(:,:,i))) + Qy
    END IF
    IF(loctype(i)==1 .OR. Fx(i)==0.0) THEN          
      axArr(:,:,i+1) = MATMUL(T,axArr(:,:,i))      
      PxArr(:,:,i+1) = MATMUL(MATMUL(T,PxArr(:,:,i)),TRANSPOSE(T)) + Qx
      KxArr(:,:,i) = 0.0
      LxArr(:,:,i) = T
    ELSE
      vx(i) = x(i)-axArr(1,1,i)
      llx = llx - (log(Fx(i)) + vx(i)*vx(i)/Fx(i))/2     
      KxArr(:,:,i) = MATMUL(MATMUL(T,PxArr(:,:,i)),TRANSPOSE(Z))/Fx(i)      
      LxArr(:,:,i) = T - MATMUL(KxArr(:,:,i),Z)
      axArr(:,:,i+1) = MATMUL(T,axArr(:,:,i)) + KxArr(:,:,i)*vx(i)      
      PxArr(:,:,i+1) = MATMUL(MATMUL(T,PxArr(:,:,i)),TRANSPOSE(LxArr(:,:,i))) + Qx
    END IF
  END DO
  
 !! BEGIN SMOOTHING LOOP!!
  DO j=N,1,-1
    IF(loctype(j)==1 .OR. Fy(j)==0.0) THEN
      ry = MATMUL(TRANSPOSE(LyArr(:,:,j)),ry)
      Ny = MATMUL(MATMUL(TRANSPOSE(LyArr(:,:,j)), Ny), LyArr(:,:,j))
    ELSE
      uy(j,:) = vy(j)/Fy(j) - RESHAPE(MATMUL(TRANSPOSE(KyArr(:,:,j)), ry), (/1/))
      My(j,:) = (1/Fy(j)) + RESHAPE(MATMUL(MATMUL(TRANSPOSE(KyArr(:,:,j)), Ny), KyArr(:,:,j)),(/1/))
      jky(j) = y(j) - uy(j,1)/My(j,1)
      ry = TRANSPOSE(Z)*vy(j)/Fy(j) + MATMUL(TRANSPOSE(LyArr(:,:,j)),ry)
      Ny = MATMUL(TRANSPOSE(Z),Z)/Fy(j) + MATMUL(MATMUL(TRANSPOSE(LyArr(:,:,j)), Ny), LyArr(:,:,j))      
    END IF
    
    IF(loctype(j)==1 .OR. Fx(j)==0.0) THEN
      rx = MATMUL(TRANSPOSE(LxArr(:,:,j)),rx)
      Nx = MATMUL(MATMUL(TRANSPOSE(LxArr(:,:,j)), Nx), LxArr(:,:,j))
    ELSE      
      ux(j,:) = vx(j)/Fx(j) - RESHAPE(MATMUL(TRANSPOSE(KxArr(:,:,j)), rx), (/1/))
      Mx(j,:) = (1/Fx(j)) + RESHAPE(MATMUL(MATMUL(TRANSPOSE(KxArr(:,:,j)), Nx), KxArr(:,:,j)), (/1/))
      jkx(j) = x(j) - ux(j,1)/Mx(j,1)
      rx = TRANSPOSE(Z)*vx(j)/Fx(j) + MATMUL(TRANSPOSE(LxArr(:,:,j)),rx)      
      Nx = MATMUL(TRANSPOSE(Z),Z)/Fx(j) + MATMUL(MATMUL(TRANSPOSE(LxArr(:,:,j)), Nx), LxArr(:,:,j))
    END IF
    predy(j,:) = RESHAPE(ayArr(:,:,j) + MATMUL(PyArr(:,:,j), ry), (/2/))
    predx(j,:) = RESHAPE(axArr(:,:,j) + MATMUL(PxArr(:,:,j), rx), (/2/))
    vary(:,:,j) = PyArr(:,:,j) - MATMUL(MATMUL(PyArr(:,:,j),Ny), PyArr(:,:,j))
    varx(:,:,j) = PxArr(:,:,j) - MATMUL(MATMUL(PxArr(:,:,j),Nx), PxArr(:,:,j))
  END DO
END SUBROUTINE crw_predict

File: crawl/R/crwPredict.R
"crwPredict" <- function(object.crwFit, predTime=NULL, speedEst=FALSE, flat=TRUE)
{
    ## Model definition/parameters ##
    data <- object.crwFit$data
    driftMod <- object.crwFit$random.drift
    stopMod <- !is.null(object.crwFit$stop.model)
    mov.mf <- object.crwFit$mov.mf
    stop.mf <- object.crwFit$stop.mf
    err.mfX <- object.crwFit$err.mfX
    err.mfY <- object.crwFit$err.mfY
    par <- object.crwFit$par
    n.errX <- object.crwFit$n.errX
    n.errY <- object.crwFit$n.errY
    n.mov <- object.crwFit$n.mov
    tn <- object.crwFit$Time.name
	if(inherits(predTime, "POSIXct")) predTime <- as.numeric(predTime)/3600

    ## Data setup ##
    if (!is.null(predTime)) {
        if(min(predTime) <  data[1, tn]) {
          warning("Predictions times given before first observation!\nOnly those after first observation will be used.")
          predTime <- predTime[predTime>=data[1,tn]]
        }
        origTime <- data[, tn]
        if (is.null(data$locType)) data$locType <- "o"
        predData <- data.frame(predTime, "p")
        names(predData) <- c(tn, "locType")
        data <- merge(data, predData,
                      by=c(tn, "locType"), all=TRUE)
        dups <- duplicated(data[, tn]) #& data[,"locType"]==1
        data <- data[!dups, ]
        mov.mf <- as.matrix(expandPred(x=mov.mf, Time=origTime, predTime=predTime))
        if (stopMod) stop.mf <- as.matrix(expandPred(x=stop.mf, Time=origTime, predTime=predTime))
        if (!is.null(err.mfX)) err.mfX <- as.matrix(expandPred(x=err.mfX, Time=origTime, predTime=predTime))
        if (!is.null(err.mfY)) err.mfY <- as.matrix(expandPred(x=err.mfY, Time=origTime, predTime=predTime))
    }
    if (object.crwFit$polar.coord) {
        lonAdjVals <- cos(round(approx(data[, tn], data[, object.crwFit$coord[2]],
                                       data[, tn])$y, 0) * pi / 180)
        if(is.na(lonAdjVals[1])) stop("Error in Longitude correction: Check to see that prediction times do not predate first observation")
        if(is.na(lonAdjVals[nrow(data)])) {
         warning('Forcasting locations past last observation: Longitude correction will be based on the last observation')
          lonAdjVals <- lonAdjVals[!is.na(lonAdjVals)][cumsum(!is.na(lonAdjVals))]
        }
    } else lonAdjVals <- rep(1, nrow(data))
    data$locType[data[,tn]%in%predTime] <- 'p'
    delta <- c(diff(data[, tn]), 1)
    a1.x <- object.crwFit$initial.state$a1.x
    P1.x <- object.crwFit$initial.state$P1.x
    a1.y <- object.crwFit$initial.state$a1.y
    P1.y <- object.crwFit$initial.state$P1.y
    y <- data[, object.crwFit$coord[2]]
    x <- data[, object.crwFit$coord[1]]
    loctype <- ifelse(is.na(x) | is.na(y), 1, 0)
    y <- ifelse(loctype == 1, 9999, y)
    x <- ifelse(loctype == 1, 9999, x)
    N <- length(loctype)
   ###
   ### Process parameters for Fortran
   ###
   if (!is.null(err.mfX)) {
      theta.errX <- par[1:n.errX]
      tau2x <- exp(2 * err.mfX %*% theta.errX)
   } else tau2x <- rep(0.0, N)
   if (!is.null(err.mfY)) {
      theta.errY <- par[(n.errX + 1):(n.errX + n.errY)]
      tau2y <- exp(2 * err.mfY %*% theta.errY)
   } else tau2y <- tau2x
   theta.mov <- par[(n.errX + n.errY + 1):(n.errX + n.errY + 2 * n.mov)]
   sig2 <- exp(2 * (mov.mf %*% theta.mov[1:n.mov]))
   b <- exp(mov.mf %*% theta.mov[(n.mov + 1):(2 * n.mov)])
   stay <- rep(0, N)
   if (stopMod) {
      stop.mf <- stop.mf
      theta.stop <- par[(n.errX + n.errY + 2 * n.mov + 1)]
      b <- b / ((1 - stop.mf) ^ exp(theta.stop))
      stay <- ifelse(b==Inf, 1, 0)
      b <- ifelse(b==Inf, 9999, b) 
   }
   if (driftMod) {
      theta.drift <- par[(n.errX + n.errY + 2 * n.mov + 1):
                                    (n.errX + n.errY + 2 * n.mov + 2)]
      b.drift <- exp(log(b) - log(1+exp(theta.drift[2])))
      sig2.drift <- exp(log(sig2) + 2 * theta.drift[1])
      call.predict <- "crwdrift_predict"
   } else {
      b.drift <- sig2.drift <- 0.0
      call.predict <- "crw_predict"
   }
                    
   movMats <- getQT(sig2, b, sig2.drift, b.drift, delta, driftMod)
   
    out <- .Fortran(call.predict,
                    tau2y=as.double(tau2y),
                    tau2x=as.double(tau2x),
                    Qmat=as.double(movMats$Qmat),
                    Tmat=as.double(movMats$Tmat),
                    x=as.double(x),
                    y=as.double(y),
                    loctype=as.integer(loctype),
                    stay=as.integer(stay),
                    ay=as.double(a1.y),
                    ax=as.double(a1.x),
                    Py=as.double(P1.y),
                    Px=as.double(P1.x),
                    lonadj=as.double(lonAdjVals),
                    N=as.integer(N),
                    lly=as.double(0),
                    llx=as.double(0),
                    My=as.double(rep(0,N)),
                    uy=as.double(rep(0,N)),
                    jky=as.double(rep(0,N)),
                    Mx=as.double(rep(0,N)),
                    ux=as.double(rep(0,N)),
                    jkx=as.double(rep(0,N)),
                    predy=as.double(matrix(0, N, 2 + driftMod)),
                    predx=as.double(matrix(0, N, 2 + driftMod)),
                    vary=as.double(array(0, c(2 + driftMod, 2 + driftMod, N))),
                    varx=as.double(array(0, c(2 + driftMod, 2 + driftMod, N))),
                    package="crawl")
                    
    predy <- data.frame(matrix(out$predy, N, 2 + driftMod))
    if (driftMod) {
        names(predy) <- c("mu.y", "theta.y", "gamma.y")
    } else names(predy) <- c("mu.y", "nu.y")
    predx <- data.frame(matrix(out$predx, N, 2 + driftMod))
    if (driftMod) {
        names(predx) <- c("mu.x", "theta.x", "gamma.x")
    } else names(predx) <- c("mu.x", "nu.x")
    vary <- zapsmall(array(out$vary, c(2 + driftMod, 2 + driftMod, N)))
    varx <- zapsmall(array(out$varx, c(2 + driftMod, 2 + driftMod, N)))
    obsFit <- data.frame(shock.x=ifelse(out$Mx!=0,out$ux/out$Mx,NA), 
                         predObs.x=ifelse(out$Mx!=0,out$jkx,NA), 
                         Vshock.x=ifelse(out$Mx!=0,1/out$Mx,NA), 
                         shock.y=ifelse(out$My!=0,out$uy/out$My,NA), 
                         predObs.y=ifelse(out$My!=0,out$jky,NA), 
                         Vshock.y=ifelse(out$My!=0,1/out$My,NA))
    obsFit$outlier.chisq <- out$ux^2/out$Mx + out$uy^2/out$My
    obsFit$naive.p.val <- 1 - pchisq(obsFit$outlier.chisq, 2)
    if (speedEst) {
        log.speed <- logSpeed(predx, predy, varx, vary, object.crwFit$polar.coord)
    } else log.speed <- NULL
    out <- list(originalData=fillCols(data), alpha.hat.y=predy, alpha.hat.x=predx,
                V.hat.y=vary, V.hat.x=varx, speed=log.speed, loglik=out$lly + out$llx)
    if (flat) {
        out <- cbind(fillCols(as.flat(out)), obsFit)
        attr(out, "flat") <- TRUE
        attr(out, "coord") <- c(x=object.crwFit$coord[1], y=object.crwFit$coord[2])
    	attr(out, "random.drift") <- driftMod
    	attr(out, "stop.model") <- object.crwFit$stopMod
    	attr(out, "polar.coord") <- object.crwFit$polar.coord
    	attr(out, "Time.name") <- tn
    } else {
       out <- append(out, list(fit.test=obsFit))
        attr(out, "flat") <- FALSE
        attr(out, "coord") <- c(x=object.crwFit$coord[1], y=object.crwFit$coord[2])
    	attr(out, "random.drift") <- driftMod
    	attr(out, "stop.model") <- object.crwFit$stopMod
    	attr(out, "polar.coord") <- object.crwFit$polar.coord
    	attr(out, "Time.name") <- tn
    }
    class(out) <- c(class(out),"crwPredict")
    return(out)
}
